proctl/variables.go

```diff
@@ -490,27 +498,71 @@ func (thread *ThreadContext) extractVariableFromEntry(entry *dwarf.Entry) (*Vari
 	return &Variable{Name: n, Type: t.String(), Value: val}, nil
 }
 
-// Extracts the value from the instructions given in the DW_AT_location entry.
-// We execute the stack program described in the DW_OP_* instruction stream, and
-// then grab the value from the other processes memory.
-func (thread *ThreadContext) extractValue(instructions []byte, addr int64, typ interface{}) (string, error) {
-	address := addr
+// Execute the stack program taking into account the current stack frame
+func (thread *ThreadContext) executeStackProgram(instructions []byte) (int64, error) {
+	regs, err := thread.Registers()
+	if err != nil {
+		return 0, err
+	}
 
-	if address == 0 {
-		regs, err := thread.Registers()
+	fde, err := thread.Process.FrameEntries.FDEForPC(regs.PC())
+	if err != nil {
+		return 0, err
+	}
+
+	fctx := fde.EstablishFrame(regs.PC())
+	cfaOffset := fctx.CFAOffset() + int64(regs.SP())
+
+	address, err := op.ExecuteStackProgram(cfaOffset, instructions)
+	if err != nil {
+		return 0, err
+	}
+	return address, nil
+}
+
+// Extracts the address of a variable, dereferencing any pointers
+func (thread *ThreadContext) extractVariableDataAddress(entry *dwarf.Entry, reader *reader.Reader) (int64, error) {
+	instructions, err := instructionsForEntry(entry)
+	if err != nil {
+		return 0, err
+	}
+
+	address, err := thread.executeStackProgram(instructions)
+	if err != nil {
+		return 0, err
+	}
+
+	// dereference pointers to get down the concrete type
+	for typeEntry, err := reader.SeekToType(entry, true, false); typeEntry != nil; typeEntry, err = reader.SeekToType(typeEntry, true, false) {
 		if err != nil {
-			return "", err
+			return 0, err
 		}
 
-		fde, err := thread.Process.FrameEntries.FDEForPC(regs.PC())
+		if typeEntry.Tag != dwarf.TagPointerType {
+			break
+		}
+
+		ptraddress := uintptr(address)
+
+		ptr, err := thread.readMemory(ptraddress, ptrsize)
 		if err != nil {
-			return "", err
+			return 0, err
 		}
+		address = int64(binary.LittleEndian.Uint64(ptr))
+	}
+
+	return address, nil
+}
 
-		fctx := fde.EstablishFrame(regs.PC())
-		cfaOffset := fctx.CFAOffset() + int64(regs.SP())
+// Extracts the value from the instructions given in the DW_AT_location entry.
+// We execute the stack program described in the DW_OP_* instruction stream, and
+// then grab the value from the other processes memory.
+func (thread *ThreadContext) extractValue(instructions []byte, addr int64, typ interface{}) (string, error) {
+	var err error
+	address := addr
```

Comment:
Might as well just use `addr` here. It isn't used after being assigned to the local, so we shouldn't have to preserve it.